028-service-binding.md

```diff
@@ -0,0 +1,415 @@
+        - name: external
+          bootstrapServers: myhost.com:443
+      ca.crt: # Strimzi cluster CA certificate
+      ca.p12: # PKCS #12 archive file for Strimzi cluster CA certificate
+      ca.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file
+    ```
+    
+**Note:** The Service Binding spec does contain some suggested secret fields for certificates, but they will not suffice to encapsulate all the certificate related information that an application needs. Hence, the suggestion of the separate fields above.
+
+Considerations:
+
+ - This approach results in a new secret being required that contains copies of existing secrets
+ - It does not help the user to decide which listener to use. The application would need to parse the secret to pick out the correct listener from the list
+
+### 3 - Application binds to KafkaUser resource that contains all details
+
+In this approach the application only binds to the `KafkaUser` and the secret created for the `KafkaUser` contains all the required details.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the secret created for the `KafkaUser` to include the bootstrap servers for each listener
+- Update the secret created for the `KafkaUser` to include the cluster certificate information
+
+Considerations:
+
+- This approach results in a lot more information being added to each `KafkaUser` secret
+- It does not help the user to decide which listener to use. The application would need to parse the secret to pick out the correct listener from the list
+
+Possible extensions to allow only one bootstrap server to be listed:
+
+ - Could only add bootstrap server addresses for listeners that support the authentication type that the `KafkaUser` supports
+ - Could update the `KafkaUser` CR to include `spec.binding.listener` field which would then determine the listener bootstrap server address that is added to the secret
+
+### 4 - Application can bind to a single custom resource
+
+In this approach the application binds to a new custom resource type, e.g. called `KafkaConnection`.
+
+Required changes:
+
+ - Add a new Custom Resource called `KafkaConnection`
+ - The `KafkaConnection` spec has several fields, cluster, listener and user, that refer to the `Kafka`, the specific listener and the `KafkaUser` that should be used
+ - The Strimzi operator adds a `status.binding` to the `KafkaConnection` status
+ - The Strimzi operator add a new secret containing all the required information
+
+Example `KafkaConnection` CR:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnection
+metadata:
+  name: barista-kafka
+spec:
+  kafka:
+    apiVersion: kafka.strimzi.io/v1beta2
+    kind: Kafka
```

Comment:
not sure if these two fields are really needed, because `KafkaConnection` refers implicitly `Kafka` resource.

Reply:
We could certainly make these optional, but the Strimzi operator does have multiple apiVersions it supports and this would mean if Strimzi has been deployed with a differently named CRD for some reason the operator can cope with that.

Reply:
I don't know ... I wonder about the usefulness of the `apiVersion` and `kind` fields as well. In particular in the relationship to the implementation. Will we use raw resources to be able to grab the information from any resource which has matching status structure? I don't think we can really do that easily. So will it end up as something what supports selected options? e.g. `kind` will always be `Kafka` and `apiVersion` will always be `kafka.strimzi.io/v1beta2` and maybe one day we add `kafka.strimzi.io/v1` as supported as well?

Reply:
I agree that the Kind seems unlikely to change, but I do wonder about the apiVersion, since as you say we may find that the `KafkaConnection` apiVersion doesn't match the `Kafka` apiVersion at some point, so allowing it to be specified means we wouldn't need to change anything at that point.